kroxylicious/src/test/java/io/kroxylicious/proxy/internal/InternalCompletableFutureTest.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.internal;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiFunction;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class InternalCompletableFutureTest {
+
+    private ExecutorService executor;
+
+    @BeforeEach
+    void beforeEach() {
+        executor = Executors.newSingleThreadExecutor();
+    }
+
+    @AfterEach
+    void afterEach() {
+        executor.shutdown();
+    }
+
+    @Test
+    void asyncChainingMethodExecutesOnThreadOfExecutor() throws Exception {
+        var threadOfExecutor = executor.submit(Thread::currentThread).get();
+        var future = InternalCompletableFuture.completedFuture(executor, null);
+
+        var actualThread = new AtomicReference<Thread>();
+        var result = future.thenAcceptAsync((u) -> {
+            assertThat(actualThread).hasValue(null);
+            actualThread.compareAndSet(null, Thread.currentThread());
```

Comment:
nit: I think we can just use set here.
```suggestion
            actualThread.set(Thread.currentThread());
```